File: src/core/GymManager.Domain/Entities/Booking.cs
using GymManager.Domain.Common;
using GymManager.Domain.Enums;

namespace GymManager.Domain.Entities;

public sealed class Booking : AuditableEntity
{
    public Guid MemberId { get; set; }
    public Guid GymHouseId { get; set; }
    public BookingType BookingType { get; set; }
    public Guid? TimeSlotId { get; set; }
    public Guid? ClassScheduleId { get; set; }
    public BookingStatus Status { get; set; }
    public DateTime BookedAt { get; set; } = DateTime.UtcNow;
    public DateTime? CheckedInAt { get; set; }
    public CheckInSource? CheckInSource { get; set; }

    public Member Member { get; set; } = null!;
    public GymHouse GymHouse { get; set; } = null!;
    public TimeSlot? TimeSlot { get; set; }
    public ClassSchedule? ClassSchedule { get; set; }

    public void CheckIn(CheckInSource source)
    {
        CheckedInAt = DateTime.UtcNow;
        CheckInSource = source;
        Status = BookingStatus.Completed;
    }

    public void Cancel()
    {
        Status = BookingStatus.Cancelled;
    }

    public void MarkNoShow()
    {
        Status = BookingStatus.NoShow;
    }

    public void Complete()
    {
        Status = BookingStatus.Completed;
    }
}

File: src/apps/gymmanager-web/src/types/booking.ts
export type BookingType = "TimeSlot" | "ClassSession";
export type BookingStatus =
  | "Confirmed"
  | "Cancelled"
  | "NoShow"
  | "Completed"
  | "WaitListed";
export type CheckInSource = "QRScan" | "ManualByStaff" | "SelfKiosk";

export interface BookingDto {
  id: string;
  memberId: string;
  gymHouseId: string;
  bookingType: number;
  timeSlotId: string | null;
  classScheduleId: string | null;
  status: number;
  bookedAt: string;
  checkedInAt: string | null;
  checkInSource: number | null;
  memberName: string;
  memberCode: string;
}

export interface TimeSlotDto {
  id: string;
  gymHouseId: string;
  date: string;
  startTime: string;
  endTime: string;
  maxCapacity: number;
  currentBookings: number;
  availableSpots: number;
}

export interface ClassScheduleDto {
  id: string;
  gymHouseId: string;
  trainerId: string;
  trainerName: string;
  className: string;
  dayOfWeek: number;
  startTime: string;
  endTime: string;
  maxCapacity: number;
  currentEnrollment: number;
  availableSpots: number;
  isRecurring: boolean;
}

export interface CreateBookingRequest {
  memberId: string;
  bookingType: number;
  timeSlotId?: string;
  classScheduleId?: string;
}

export interface CreateTimeSlotRequest {
  date: string;
  startTime: string;
  endTime: string;
  maxCapacity: number;
}

export interface CreateClassScheduleRequest {
  trainerId: string;
  className: string;
  dayOfWeek: number;
  startTime: string;
  endTime: string;
  maxCapacity: number;
  isRecurring: boolean;
}

export interface UpdateClassScheduleRequest {
  className: string;
  dayOfWeek: number;
  startTime: string;
  endTime: string;
  maxCapacity: number;
  isRecurring: boolean;
}

export interface CheckInRequest {
  source: number;
}
